expression/simple_rewriter.go

```diff
@@ -245,6 +247,33 @@ func (sr *simpleRewriter) Leave(originInNode ast.Node) (retNode ast.Node, ok boo
 			Value:   types.NewStringDatum(v.Selector.String()),
 			RetType: types.NewFieldType(mysql.TypeVarchar),
 		})
+	case *ast.SetCollationExpr:
+		arg := sr.pop()
+		if collate.NewCollationEnabled() {
+			var collInfo *charset.Collation
+			// TODO(bb7133): use charset.ValidCharsetAndCollation when its bug is fixed.
+			if collInfo, sr.err = charset.GetCollationByName(v.Collate); sr.err != nil {
+				sr.err = charset.ErrUnknownCollation.GenWithStackByArgs(v.Collate)
+				break
+			}
+			chs := arg.GetType().Charset
+			if chs != "" && collInfo.CharsetName != chs {
+				sr.err = charset.ErrCollationCharsetMismatch.GenWithStackByArgs(collInfo.Name, chs)
+				break
+			}
+		}
+		// SetCollationExpr sets the collation explicitly, even when the evaluation type of the expression is non-string.
+		if _, ok := arg.(*Column); ok {
+			// Wrap a cast here to avoid changing the original FieldType of the column expression.
+			casted := BuildCastFunction(sr.ctx, arg, arg.GetType())
+			casted.GetType().Collate = v.Collate
```

Comment:
Should we create a new type first and then cast the arg to this expected type?
```
exprType := arg.GetType().Clone()
exprType.Collate = v.Collate
casted := BuildCastFunction(sr.ctx, arg, exprType)
```

Reply:
Yeah